tcp/server/server.go

```diff
@@ -0,0 +1,229 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/plgd-dev/go-coap/v3/message"
+	"github.com/plgd-dev/go-coap/v3/message/pool"
+	coapNet "github.com/plgd-dev/go-coap/v3/net"
+	"github.com/plgd-dev/go-coap/v3/net/blockwise"
+	"github.com/plgd-dev/go-coap/v3/net/monitor/inactivity"
+	"github.com/plgd-dev/go-coap/v3/pkg/connections"
+	"github.com/plgd-dev/go-coap/v3/tcp/client"
+)
+
+// A ServerOption sets options such as credentials, codec and keepalive parameters, etc.
+type ServerOption interface {
+	TCPServerApply(cfg *Config)
+}
+
+// Listener defined used by coap
+type Listener interface {
+	Close() error
+	AcceptWithContext(ctx context.Context) (net.Conn, error)
+}
+
+type Server struct {
+	listenMutex sync.Mutex
+	listen      Listener
+	ctx         context.Context
+	cancel      context.CancelFunc
+	cfg         *Config
+}
+
+func New(opt ...ServerOption) *Server {
+	cfg := DefaultConfig
+	for _, o := range opt {
+		o.TCPServerApply(&cfg)
+	}
+
+	ctx, cancel := context.WithCancel(cfg.Ctx)
+
+	if cfg.CreateInactivityMonitor == nil {
+		cfg.CreateInactivityMonitor = func() client.InactivityMonitor {
+			return inactivity.NewNilMonitor[*client.ClientConn]()
+		}
+	}
+	if cfg.MessagePool == nil {
+		cfg.MessagePool = pool.New(0, 0)
+	}
+
+	if cfg.Errors == nil {
+		cfg.Errors = func(error) {
+			// default no-op
+		}
+	}
+	if cfg.GetToken == nil {
+		cfg.GetToken = message.GetToken
+	}
+	errorsFunc := cfg.Errors
+	// assign updated func to opts.errors so opts.handler also uses the updated error handler
+	cfg.Errors = func(err error) {
+		if errors.Is(err, context.Canceled) || errors.Is(err, io.EOF) || strings.Contains(err.Error(), "use of closed network connection") {
```

Comment:
we have `net.IsCancelOrCloseError` for this check, should be substituted for all remaining code with 

```go
strings.Contains(err.Error(), "use of closed network connection"
```

Reply:
fixed